src/System.Private.CoreLib/src/System/Array.Portable.cs

```diff
@@ -0,0 +1,1659 @@
+        // IndexOf is that it is supposed to use IEquatable<T> if possible, but that requirement
+        // cannot be expressed in IL directly due to the lack of constraints.
+        // Instead, specialization at call time is used within the compiler. 
+        // 
+        // General Approach
+        // - Perform fancy redirection for Array.GetComparerForReferenceTypesOnly<T>(). If T is a reference 
+        //   type or UniversalCanon, have this redirect to EqualityComparer<T>.get_Default, Otherwise, use 
+        //   the function as is. (will return null in that case)
+        // - Change the contents of the IndexOf functions to have a pair of loops. One for if 
+        //   GetComparerForReferenceTypesOnly returns null, and one for when it does not. 
+        //   - If it does not return null, call the EqualityComparer<T> code.
+        //   - If it does return null, use a special function StructOnlyEquals<T>(). 
+        //     - Calls to that function result in calls to a pair of helper function in 
+        //       EqualityComparerHelpers (StructOnlyEqualsIEquatable, or StructOnlyEqualsNullable) 
+        //       depending on whether or not they are the right function to call.
+        // - The end result is that in optimized builds, we have the same single function compiled size 
+        //   characteristics that the old EqualsOnlyComparer<T>.Equals function had, but we maintain 
+        //   correctness as well.
+        private static EqualityComparer<T> GetComparerForReferenceTypesOnly<T>()
+        {
+#if !CORERT
+            return System.Runtime.CompilerServices.RuntimeHelpers.IsReferenceOrContainsReferences<T> () ? EqualityComparer<T>.Default : null;
+#else
+            return EqualityComparer<T>.Default;
+#endif
+        }
+
+        private static bool StructOnlyEquals<T>(T left, T right)
+        {
+            return left.Equals(right);
+        }
+
+        private sealed class ArrayEnumerator : IEnumerator, ICloneable
+        {
+            private Array _array;
+            private int _index;
+            private int _endIndex; // cache array length, since it's a little slow.
+
+            internal ArrayEnumerator(Array array)
+            {
+                _array = array;
+                _index = -1;
+                _endIndex = array.Length;
+            }
+
+            public bool MoveNext()
+            {
+                if (_index < _endIndex)
+                {
+                    _index++;
+                    return (_index < _endIndex);
+                }
+                return false;
+            }
+
+            public Object Current
+            {
+                get
+                {
+                    if (_index < 0) throw new InvalidOperationException(SR.InvalidOperation_EnumNotStarted);
+                    if (_index >= _endIndex) throw new InvalidOperationException(SR.InvalidOperation_EnumEnded);
+                    return _array.GetValueWithFlattenedIndex_NoErrorCheck(_index);
```

Comment:
ArrayEnumerator differs from CoreCLR at this place (we avoid a bunch of checks this way), but it makes the enumerator rather non-portable (see my pending change to it in #3192).